### PR TITLE
Limit time step based on nucleation probability

### LINKED ISSUE #12168: Limit simulation timestep based on nucleation probability distribution
## Rationale

The Discrete Nucleation system samples a Poisson distribution to determine nucleation events. This works only as long as the probability for an more than event to occur in the current timestep at a given point is low (i.e. if either zero or one event are likely to occur).

## Description

The inserter will determine the largest nucleation rate over all quadrature points in the simulation and set the upper timestep limit to keep the resulting maximal probability below a given value.

Due to execution order issues this determination is made at the end of a timestep before adaptivity occurs. As mesh coarsening will increase nucleation rates (as the rate density is integrated over larger volumes in the coarsened elements) a modification will have to be made to the `DiscreteNucleationMarker` to avoid coarsening of high rate regions.

## Impact

Improved reliability of the nucleation system.

### PR DESCRIPTION
**Pull request by dschwen**
This limits the time step based on the overall nucleation rate. The user specifies an upper limit to the probability of _two or more nuclei_ to appear during a single timestep. This limits the timestep and puts it in the rare event regime where we are sampling the poisson statistic correctly (and avoid overlapping nuclei to form).

Closes #12168

### CODE REVIEW

**Review by aeslaughter — COMMENTED** [review]
I approve the python changes. Thanks.

**Comment by dschwen** [comment]
@aeslaughter you wouldn't mind adding SQA tags to this python tests you touched, would you?

**Comment by dschwen** [comment]
Ok, I've tried to add some, but I cannot fill out the `design =` tags.

File: modules/phase_field/doc/content/source/postprocessors/DiscreteNucleationData.md [hunk 1]
```diff
@@ -4,9 +4,10 @@
 
 Outputs the number of currently active nucleation sites maintained by the given
 [DiscreteNucleationInserter](/DiscreteNucleationInserter.md) (`value` option
-`COUNT`) or a boolean (0/1) value denoting if at least one new nucleus was
+`COUNT`), a boolean (0/1) value denoting if at least one new nucleus was
 inserted or removed (i.e. its hold time expired) for the current timestep
-(`value` option `UPDATE`).
+(`value` option `UPDATE`), or the total nucleation rate for the entire domain
+(`value` option `RATE`).
```
Comment (amjokisaari) [inline modules/phase_field/doc/content/source/postprocessors/DiscreteNucleationData.md]: The wording here is a bit confusing.  I'd suggest something like "Outputs one of three possible types of data: 1) xyz, 2) ... 3)... "

File: modules/phase_field/include/postprocessors/DiscreteNucleationTimeStep.h [hunk 2]
```diff
@@ -38,6 +38,13 @@ class DiscreteNucleationTimeStep : public GeneralPostprocessor
 
   /// User specified nucleation time step
   const Real _dt_nucleation;
+
+  /**
+   * Maximum total event expectation value that is low enough so that the
+   * probability for more than one * nucleation event to occurr in a single
```
Comment (amjokisaari) [inline modules/phase_field/include/postprocessors/DiscreteNucleationTimeStep.h]: occurr -> occur

File: modules/phase_field/src/postprocessors/DiscreteNucleationData.C [hunk 3]
```diff
@@ -18,11 +18,12 @@ validParams<DiscreteNucleationData>()
...
   params.addClassDescription("Output diagnostic data on a DiscreteNucleationInserter");
   params.addRequiredParam<UserObjectName>("inserter", "DiscreteNucleationInserter user object");
-  MooseEnum valueType("COUNT UPDATE", "COUNT");
+  MooseEnum valueType("COUNT UPDATE RATE", "COUNT");
   params.addRequiredParam<MooseEnum>(
       "value",
       valueType,
-      "Select to output number of active nuclei or wether a change to the nucleus list occurred");
+      "Select to output number of active nuclei, wether a change to "
+      "the nucleus list occurred, or the total rate over the entire domain");
```
Comment (amjokisaari) [inline modules/phase_field/src/postprocessors/DiscreteNucleationData.C]: this wording is confusing for the second information in the list.

File: modules/phase_field/src/postprocessors/DiscreteNucleationTimeStep.C [hunk 4]
```diff
@@ -20,7 +21,15 @@ validParams<DiscreteNucleationTimeStep>()
...
       "Return a timestep limit for nucleation event to be used by IterationAdaptiveDT");
   params.addRequiredParam<Real>("dt_max",
-                                "Timestep to cut back to at the start of sa nucleation event");
+                                "Time step to cut back to at the start of a nucleation event");
+  params.addRangeCheckedParam<Real>(
+      "p2nucleus",
+      0.01,
+      "p2nucleus > 0 & p2nucleus < 1",
+      "Maximum probability for more than one nucleus to appear during a time "
+      "step. This will limit the time step base on the total nucleation rate for "
```
Comment (amjokisaari) [inline modules/phase_field/src/postprocessors/DiscreteNucleationTimeStep.C]: base -> based

**Review by amjokisaari — COMMENTED** [review]
Is there documentation (for the website) that describes the theory and algorithm for the global time step limitation calculation?

Does the time step now have to be limited by picking some probability (is the user forced to go through this step) or not?  As written, it looks like it must be done.

**Comment by dschwen** [comment]
> Is there documentation (for the website) that describes the theory and algorithm for the global time step limitation calculation?

There is this:
http://mooseframework.inl.gov/docs/PRs/12183/site/source/postprocessors/DiscreteNucleationTimeStep.html

> Does the time step now have to be limited by picking some probability [..]

Yeah, but there is a default value of 1% in the code.

**Comment by friedmud** [comment]
Really cool!

**Review by permcody — APPROVED** [review]

**Pull request merged**